Use full size image for lightbox effect when thumbnail is used for image.

dhaval-parekh opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6898 

Changes in PR adds the srcset attribute to `<img/>` or `<amp-img/>` tag that has lightbox effect enabled in order to show full size image in lightbox.

**Gallery**

https://user-images.githubusercontent.com/8168027/163762934-d8c1a3a4-e941-4a9d-818c-63845a75f18a.mov

**Image**

https://user-images.githubusercontent.com/8168027/163763072-d3b01aad-4a11-4a43-bb52-bc48b25fbde8.mov


## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

dhaval-parekh committed:
Enabled lightbox for image and gallery with thumbnail size image

dhaval-parekh committed:
Add test cases

milindmore22 commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -394,6 +394,28 @@ private function adjust_and_replace_nodes_in_array_map( $node_lists ) {
+			! $node->hasAttribute( Attribute::SRCSET )
+			&&
+			$node->hasAttribute( 'data-amp-lightbox' )
+			&&
```
This should be OR condition? 
If the user opt-out for Lightbox but opt-in for carousel it should also generate srcset otherwise it will show distorted images in the carousel

milindmore22 replied:
https://user-images.githubusercontent.com/1132541/163835692-f2cb8198-457e-4812-a18c-89dfc937fd99.mov

dhaval-parekh replied:
Here we are setting `srcset` because we need to show full size image in lightbox.

milindmore22 reviewed (changes requested):
minor change: carousel using thumbnail should also get generate srcset

milindmore22 reviewed (approved):
LGTM 👍🏼

westonruter commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -394,6 +394,28 @@ private function adjust_and_replace_nodes_in_array_map( $node_lists ) {
+		if (
+			! $node->hasAttribute( Attribute::SRCSET )
+			&&
+			$node->hasAttribute( 'data-amp-lightbox' )
```
Sometimes the `data-amp-lightbox` attribute can be on an ancestor node:

https://github.com/ampproject/amp-wp/blob/49a8eef34942d9c4429bf61be3c949d986fcc7bf/includes/embeds/class-amp-gallery-embed-handler.php#L101-L111

https://github.com/ampproject/amp-wp/blob/a41c2f682caccd6bd67e364dffdfdb1547c07e65/includes/sanitizers/class-amp-gallery-block-sanitizer.php#L80-L85

https://github.com/ampproject/amp-wp/blob/1916a9129c449d13d0072c30a37bac61753ac202/includes/sanitizers/class-amp-img-sanitizer.php#L540-L541

Or will that have always been copied to the `amp-img` at this point?

dhaval-parekh replied:
At this point it's always copied to child img or amp-img because of this.


https://github.com/ampproject/amp-wp/blob/2709f5f63f74efc893248e904b10051948ba9c68/includes/embeds/class-amp-core-block-handler.php#L283-L308

westonruter replied:
What about a `gallery` shortcode?

dhaval-parekh replied:
Ohh. Right, `gallery` shortcode does add `data-amp-lightbox` to parent.

However, I have remove the condition of checking `data-amp-lightbox` attribute in [this](https://github.com/ampproject/amp-wp/pull/7061/commits/d66a48413179c2aef961e08d53eb8c24456d2722) commit.

dhaval-parekh committed:
Remove lightbox condition to set srcset to image

westonruter reviewed (changes requested):
I've tried with `post_content` that includes a single Image, a Block Gallery, and a `gallery` shortcode:

```html
<!-- wp:heading -->
<h2>Image</h2>
<!-- /wp:heading -->

<!-- wp:image {"id":2784,"sizeSlug":"thumbnail","linkDestination":"none","ampLightbox":true} -->
<figure class="wp-block-image size-thumbnail"><img src="https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-150x150.jpg" alt="" class="wp-image-2784"/></figure>
<!-- /wp:image -->

<!-- wp:heading -->
<h2>Gallery Block</h2>
<!-- /wp:heading -->

<!-- wp:gallery {"linkTo":"none","ampLightbox":true} -->
<figure class="wp-block-gallery has-nested-images columns-default is-cropped"><!-- wp:image {"id":3004,"sizeSlug":"thumbnail","linkDestination":"none"} -->
<figure class="wp-block-image size-thumbnail"><img src="https://wordpressdev.lndo.site/content/uploads/2021/05/bison-web-stories-poster-150x150.jpg" alt="" class="wp-image-3004"/><figcaption>Bison</figcaption></figure>
<!-- /wp:image -->

<!-- wp:image {"id":2784,"sizeSlug":"thumbnail","linkDestination":"none"} -->
<figure class="wp-block-image size-thumbnail"><img src="https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-150x150.jpg" alt="" class="wp-image-2784"/></figure>
<!-- /wp:image --></figure>
<!-- /wp:gallery -->

<!-- wp:heading -->
<h2>Gallery Shortcode</h2>
<!-- /wp:heading -->

<!-- wp:shortcode -->
[gallery ids="3004,2784" size="thumbnail" amp-lightbox="true"]
<!-- /wp:shortcode -->
```

For images generated by a `gallery` shortcode, there is no `data-id` attribute being added to them, so no `srcset` is being added.

But more problematically, I'm noticing an unexpected side effect of adding `srcset`: on larger screen resolutions it is causing the higher resolution images to be supplied undesirably:

<img width="949" alt="image" src="https://user-images.githubusercontent.com/134745/164869303-911bd800-7a5c-43a4-876d-2d6c70c94abf.png">

If there isn't a way to prevent this, then I don't know if this solution is going to work. 🙍‍♂️

dhaval-parekh committed:
Add srcset attribute to gallery image render from shortcode

dhaval-parekh committed:
Add sizes attribute to image

dhaval-parekh committed:
Update unit test cases

dhaval-parekh committed:
Update logic to set sizes to image

dhaval-parekh commented:
> For images generated by a gallery shortcode, there is no data-id attribute being added to them, so no srcset is being added.

I have added the necessary attributes to the image from the gallery generated by shortcode.


> But more problematically, I'm noticing an unexpected side effect of adding srcset: on larger screen resolutions it is causing the higher resolution images to be supplied undesirably:

Thank you for noticing that issue. I was able to reproduce it. And I think it's because sizes attributes were missing. I have added it.

https://user-images.githubusercontent.com/8168027/165497502-a44a182b-7955-4017-9881-77dc68854fdb.mov

dhaval-parekh committed:
Change the value of sizes for gallery shortcode images

dhaval-parekh committed:
Fix unit test cases

westonruter commented:
Even with the `sizes` attribute added, I'm not seeing a difference. On Desktop, a larger image is being used than the thumbnail size:

Non-AMP | AMP
--------|------
![image](https://user-images.githubusercontent.com/134745/165629676-58099a49-5566-4fbf-8547-fbb014823634.png) | ![image](https://user-images.githubusercontent.com/134745/165629740-903af60a-ee74-4968-9b86-8914ace15c37.png)

The generated `amp-img` tag is:

```html
<amp-img width="150" height="150" data-id="2784" src="https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-150x150.jpg" alt="" class="wp-image-2784 amp-wp-enforced-sizes" data-amp-lightbox="" lightbox="" layout="intrinsic" srcset="https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-scaled.jpg 2560w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-300x196.jpg 300w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-1024x668.jpg 1024w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-768x501.jpg 768w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-1536x1002.jpg 1536w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-2048x1336.jpg 2048w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-1200x783.jpg 1200w, https://wordpressdev.lndo.site/content/uploads/2021/04/American_bison_k5680-1-1980x1291.jpg 1980w" sizes="(max-width: 150px) 100vw, 150px">
```

I thought maybe this could be fixed by changing the `sizes` attribute from `(max-width: 150px) 100vw, 150px` to just `150px`, but I didn't see any difference. Granted I'm not an expert on responsive images and `srcset`/`sizes` always confuse me.

Here's a minimal test case:

```html
<!DOCTYPE html>
<html ⚡>
<head>
	<meta charset="utf-8" />
	<title>My AMP Page</title>
	<link rel="canonical" href="self.html" />
	<meta name="viewport" content="width=device-width" />
	<style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
	<script async src="https://cdn.ampproject.org/v0.js"></script>
	<script src="https://cdn.ampproject.org/v0/amp-lightbox-0.1.js" async="" custom-element="amp-lightbox"></script>
	<script src="https://cdn.ampproject.org/v0/amp-lightbox-gallery-0.1.js" async="" custom-element="amp-lightbox-gallery"></script>
</head>
<body>
<amp-img
		width="150"
		height="150"
		data-id="2784"
		src="https://via.placeholder.com/150x150"
		alt=""
		class="wp-image-2784 amp-wp-enforced-sizes"
		data-amp-lightbox=""
		lightbox=""
		layout="intrinsic"
		srcset="
			https://via.placeholder.com/2560x1670 2560w, 
			https://via.placeholder.com/300x196 300w, 
			https://via.placeholder.com/1024x668 1024w, 
			https://via.placeholder.com/768x501 768w, 
			https://via.placeholder.com/1536x1002 1536w, 
			https://via.placeholder.com/2048x1336 2048w, 
			https://via.placeholder.com/1200x783 1200w, 
			https://via.placeholder.com/1980x1291 1980w
		"
		sizes="(max-width: 150px) 100vw, 150px"
></amp-img>
</body>
</html>
```

You can see that the desired 150x150 image is not being loaded:

<img src="https://user-images.githubusercontent.com/134745/165632744-1862694e-8de5-4d96-95e4-a9f2acf87838.png" width=150 height=150>

dhaval-parekh committed:
Merge branch 'develop' into bug/6898-gallery-block-lightbox

dhaval-parekh committed:
Add thumbnail size to image srcset and update value of sizes attribute

dhaval-parekh committed:
Update unit test case

westonruter commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -394,6 +394,45 @@ private function adjust_and_replace_nodes_in_array_map( $node_lists ) {
+			}
+
+			$thumbnail_image = wp_get_attachment_image_url( $attachment_id );
+			$srcset          = $thumbnail_image . ' 150w, ' . $srcset;
```
The number 150 shouldn't be hard-coded since the value can be changed on the Media screen:

![Screen Shot 2022-05-02 at 10 11 02](https://user-images.githubusercontent.com/134745/166293093-8b9fa329-5412-44ee-967a-71b3206e5d42.png)

dhaval-parekh replied:
Thank you for noticing. I have addressed this in my latest commit.

dhaval-parekh committed:
Merge branch 'develop' into bug/6898-gallery-block-lightbox

dhaval-parekh committed:
Use thumbnail size set in setting page for srcset

dhaval-parekh commented on includes/sanitizers/class-amp-img-sanitizer.php:
```diff
@@ -394,6 +394,46 @@ private function adjust_and_replace_nodes_in_array_map( $node_lists ) {
+				$size = ! empty( $matches['size'] ) ? $matches['size'] : 'thumbnail';
+			}
+
+			$size  = ( 'thumbnail' === $size ) ? [ absint( $thumbnail_size / 3 ) ] : $size;
```
Using 1/3 of the width, So it also accounts for devices that have higher pixel density.

dhaval-parekh commented:
I have made changes to prevent higher resolution images from loading.

However, with current changes, there is one drawback. That, If a bigger image is used in the content and the same image is used as a thumbnail then it will still use a higher resolution image that is already loaded.


<img width="1676" alt="image" src="https://user-images.githubusercontent.com/8168027/166492153-2155534d-95c8-4348-862e-9a8709e772c2.png">

dhaval-parekh committed:
Remove non-existing function from @cover and @use

dhaval-parekh committed:
Try to fix phpunit test case

dhaval-parekh committed:
Update unit test cases

dhaval-parekh committed:
Merge branch 'develop' into bug/6898-gallery-block-lightbox
